include/Dialects/LinalgExt/LinalgExtOps.td

```diff
@@ -191,4 +191,129 @@ def LinalgExt_TileYieldOp : LinalgExt_PureOp<"tile_yield", [
+    The only allowed terminator is `linalg_ext.perform_concurrently`, which dictates
+    how the results of all parallel invocations should be reconciled into a full
+    value that will be returned from `in_parallel`. Multi-value returns are encoded
+    by including multiple operations inside the `perform_concurrently` block.
+  }];
+  let arguments = (ins Index:$num_threads);
+
+  let results = (outs Variadic<AnyType>:$results);
+  let regions = (region SizedRegion<1>:$region);
+
+  let printer = [{ return ::print(p, *this); }];
+  let parser = [{ return ::parse$cppClass(parser, result); }];
+  let verifier = [{ return ::verify(*this); }];
+
+  let builders = [
+    OpBuilder<(ins "Value":$num_threads,
+      CArg<"function_ref<void(OpBuilder &, Location, Value)>", "nullptr">)>
+  ];
+  let extraClassDeclaration = [{
+    Value getThreadIndex() { return getBody()->getArgument(0); }
+    static void ensureTerminator(Region &region, Builder &builder, Location loc);
+  }];
+}
+
+def LinalgExt_PerformConcurrentlyOp : LinalgExt_PureOp<"perform_concurrently", [
+       NoSideEffect,
+       Terminator,
+       SingleBlockImplicitTerminator<"linalg_ext::EndPerformConcurrentlyOp">,
+      ]> {
+  let summary = "terminates a `in_parallel` block";
+  let description = [{
+    `linalg_ext.perform_concurrently` is a designated terminator for the blocks
+    of `linalg_ext.in_parallel` operations. The terminator contains a single block
+    itself, which describes how the results of each parallel invocation are to be
+    reconciled into a single value to be returned from the parallel invocation.
+    One operation in this terminator's block corresponds to a single return of
+    `in_parallel`.
+  }];
+
+  let regions = (region SizedRegion<1>:$region);
+
+  let printer = [{ return ::print(p, *this); }];
+  let parser = [{ return ::parse$cppClass(parser, result); }];
+  let verifier = [{ return ::verify(*this); }];
+
+  // TODO(apaszke, ntv): Add an interface for ops that can appear inside
+  // perform_concurrently.
+  let extraClassDeclaration = [{
+    SmallVector<Type> resultTypes() {
+      SmallVector<Type> ret;
+      for (Operation& op : *getBody()) {
+        if (auto sliceOp = llvm::dyn_cast<ParallelInsertSliceOp>(op)) {
+          ret.push_back(sliceOp.resultType());
+          continue;
+        }
+        if (auto endPerformOp = llvm::dyn_cast<EndPerformConcurrentlyOp>(op)) {
+          continue;
+        }
+        llvm_unreachable("Unexpected operation in perform_concurrently");
+      }
+      return ret;
+    };
+  }];
+}
+
+def LinalgExt_EndPerformConcurrentlyOp : LinalgExt_PureOp<"end_perform_concurrently", [
+       NoSideEffect, Terminator]> {
+  let summary = "terminates a `perform_concurrently` block";
+  let description = [{
+    A designated terminator for `perform_concurrently`. It's not expected to appear
+    in the textual form of the IR.
+  }];
+}
+
+// Note that we cannot mark this operation as pure (NoSideEffects), even
```

Comment:
I'd even that this in the actual op description, this is part of the semantics.

Reply:
Done.